Fix RW window for accessing full flash

SergiiDmytruk opened the pull request:
* Make several attempts to enable full flash access in `enable_capsule_smi()`.  This prevents a situation when capsule update fails right away due to `EFI_UNSUPPORTED` on first attempt to communicate with SMMSTORE.
* Bump EDK to get a related fix there.
* Fix RW window in fast SPI flash code.

This is part of fixing https://github.com/Dasharo/dasharo-issues/issues/1338.

EDK part: https://github.com/Dasharo/edk2/pull/255

*ref: ncm-1850*

SergiiDmytruk committed:
drivers/efi/capsules.c: try to enable full flash access multiple times

enable_capsule_smi() is responsible for enabling the access by sending a
command to SMMSTORE, but there seems to be something related to Clevo
laptops that makes SMIs prone to be lost and unhandled.

The implementation is modeled after EDK2 code which uses SMMSTORE in
https://github.com/tianocore/edk2/commit/034de59fb7a0048525109eeaba76d78d48b21ef5
with a difference that up to 11 attempts in total are made and the delay
is not done via a CPU loop.

Upstream-Status: Pending
Change-Id: I95d61174806988137869d37335fa4f4f4a052348
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
payloads/edk2/Kconfig.dasharo: bump EDK

For better logs and more reliable SMIs.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I3da3b7640ad42451fad1271d95a71695807b07f9
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

philipanda commented on src/drivers/efi/capsules.c:
```diff
@@ -792,8 +793,20 @@ static void enable_capsule_smi(void *unused)
-		       (void*)(uintptr_t)uefi_capsule_count);
+	/* SMM can occasionally ignore SMI calls, so retry several times on failure. */
+	uint8_t retries_left = 10;
+	while (1) {
```
Purely visual suggestion, but I find it weird to use an infinite loop when the number of iterations is clearly capped by `retries_left`.

Why not do
```c
while (retries_left != 0)
```
and replace the lines
 ```c
if (retries_left-- == 0)
  break;
```
with
```c
retries_left--;
```

SergiiDmytruk replied:
Then it will be number of tries and the last attempt will be followed by a delay.  Although the condition can be checked as you suggest while omitting the delay, can do that.

philipanda replied:
Right, I've missed that. In that case the first call of `call_smm` could be moved in front of the loop, and then the loop could first check the return value, then wait the delay, and retry at the end of an iteration. Anyway, I don't think it's important and only subjective

philipanda reviewed (dismissed):
Only an insignificant subjective style suggestion

SergiiDmytruk reviewed (commented):
This PR doesn't fix the issue as intended, so marking as a draft for now.

krystian-hebel commented:
Not much useful info could be obtained from capsule update process, it just hanged somewhere inside SMI handler, after printing input parameters, already on the first read. They looked mostly OK, but reported `param` address _could_ be in different than it should be. The address was too convincing to be disregarded as EC debugger garbage. I don't have the logs from that though, it was obtained on @mkopec's laptop. We could try adding more debug output to `rdev_readat()` or even raw SPI driver calls to try to pinpoint where exactly it hangs, but this is time consuming.

I do have log from OS booted normally (dirty because I also did https://review.coreboot.org/c/coreboot/+/86038, didn't help):
[nvc.log](https://github.com/user-attachments/files/21127275/nvc.log)

Interesting points:
- There are 60 `SMI_STS: APM` lines, 54 of them are preceded by double (!?) empty `SMI_STS: TCO` entries.
- In another entry, just before those 54, between two TCOs there are logs from SMMSTORE search and nothing else, so it should be counted as 55th set of TCO -> TCO -> APM entries.
- According to https://github.com/Dasharo/coreboot/blob/e2e2acc0a045f283e24de04743a85e3d8fbf1c23/src/soc/intel/common/block/smm/smihandler.c#L443, one TCO could exist, no idea where the second one comes from.
- Something strange happens along the way, SMMSTORE is reported to be cleared 5 times times during boot, in some cases a write happens immediately after. It this is normal, I'm surprised we haven't wore out flash yet. Example:

<details>

```
[NOTE ]  coreboot-dell_optiplex_7010_9010_v0.1.2-rc1-133-gaa3639ed8cec-dirty-v1.0.0-rc3 Sun Jul 06 13:45:35 UTC 2025 x86_32 smm starting (log level: 0)...
[DEBUG]  SMI_STS: TCO 


[NOTE ]  coreboot-dell_optiplex_7010_9010_v0.1.2-rc1-133-gaa3639ed8cec-dirty-v1.0.0-rc3 Sun Jul 06 13:45:35 UTC 2025 x86_32 smm starting (log level: 0)...
[DEBUG]  SMI_STS: TCO 


[NOTE ]  coreboot-dell_optiplex_7010_9010_v0.1.2-rc1-133-gaa3639ed8cec-dirty-v1.0.0-rc3 Sun Jul 06 13:45:35 UTC 2025 x86_32 smm starting (log level: 0)...
[DEBUG]  SMI_STS: APM 
[DEBUG]  Raw clear SMM store, param = 0x493d6000
[DEBUG]  FMAP: area SMMSTORE found @ 1020000 (262144 bytes)


[NOTE ]  coreboot-dell_optiplex_7010_9010_v0.1.2-rc1-133-gaa3639ed8cec-dirty-v1.0.0-rc3 Sun Jul 06 13:45:35 UTC 2025 x86_32 smm starting (log level: 0)...
[DEBUG]  SMI_STS: TCO 
[WARN ]  EFIVARS: No Firmware Volume header present
[WARN ]  EFIVARS: Failed to validate firmware header
[WARN ]  EFIVARS: No Firmware Volume header present
[WARN ]  EFIVARS: Failed to validate firmware header


[NOTE ]  coreboot-dell_optiplex_7010_9010_v0.1.2-rc1-133-gaa3639ed8cec-dirty-v1.0.0-rc3 Sun Jul 06 13:45:35 UTC 2025 x86_32 smm starting (log level: 0)...
[DEBUG]  SMI_STS: TCO 
[WARN ]  EFIVARS: No Firmware Volume header present
[WARN ]  EFIVARS: Failed to validate firmware header
[WARN ]  EFIVARS: No Firmware Volume header present
[WARN ]  EFIVARS: Failed to validate firmware header


[NOTE ]  coreboot-dell_optiplex_7010_9010_v0.1.2-rc1-133-gaa3639ed8cec-dirty-v1.0.0-rc3 Sun Jul 06 13:45:35 UTC 2025 x86_32 smm starting (log level: 0)...
[DEBUG]  SMI_STS: APM 
[DEBUG]  Raw write to SMM store, param = 0x493d6000
[DEBUG]  FMAP: area SMMSTORE found @ 1020000 (262144 bytes)
[DEBUG]  smm store: writing 0x49824000 block 0, offset=0x0, size=10000
```

</details>

SergiiDmytruk commented:
> Not much useful info could be obtained from capsule update process, it just hanged somewhere inside SMI handler, after printing input parameters, already on the first read.

Is that a side-effect of debugging SMM?  Previously Novacustom laptops hanged only with debug changes to IFD/FMAP.

As a workaround I was thinking of changing capsule update in EDK to read FMAP first and then ignore flash read failures unless they touch SI_BIOS/GBE/etc. But if hangs are common for some reason now, this won't work.

krystian-hebel commented:
> Is that a side-effect of debugging SMM?

I hope not, but at this point it is possible. It doesn't seem to return to UEFI, last printed line was `FMAP: area SMMSTORE found` (or the one before that? can't remember), and not `smm store: reading ...`.

I think I'll just write an app for sending raw SMMSTORE commands, maybe it will at least make reproduction easier so that we won't have to flash new image every time.

SergiiDmytruk commented:
> I think I'll just write an app for sending raw SMMSTORE commands, maybe it will at least make reproduction easier so that we won't have to flash new image every time.

I just make the update code fail after reading the firmware to not worry about the update, but maybe a separate app won't trash cbmem log buffer.

krystian-hebel commented:
I was able to generate SMI from Linux userspace, and first observation is that it isn't TCO -> TCO -> APM, but APM -> TCO -> TCO, and TCO isn't generated for APM other than SMMSTORE. It may be possible that TCO bit is set even if SPI transaction is initiated from within SMM.

krystian-hebel commented:
Tool for doing raw SMMSTORE operations: https://github.com/Dasharo/coreboot/pull/713

In order to use it with full flash access, we would have to disable capsule handling in edk2.

SergiiDmytruk committed:
soc/intel/fast_spi/mmap_boot.c: fix adding RW decode window

When RW decode window was added without extended MMAP window,
xlate_region_device_ro_init() was passed an array with uninitialized
second element (corresponds to the extended window) because
real_dev_windows[] was filled by indices rather than by adding the next
element in the next free slot.

Clearly separate storage of windows from array passed to
xlate_region_device_ro_init() and fill that array dynamically with
copies of windows (can't discard originals at least for MMAP windows as
they are used by other functions, preserving RW window for consistency).

This must be squashed with the previous changes to this file on the next
rebase.

Upstream-Status: Pending
Change-Id: Ib08c01d3f13dafec76ea6df20ac30cf505e16c4e
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
util/lint/kconfig_lint: fix negation of a match

This fixes the following Perl error (which prevents pre-commit hook from
succeeding):

    Possible precedence problem between ! and pattern match (m//)

`!$filename =~ /~$/` was apparently interpreted as
`(!$filename) =~ /~$/`.  Fix that by wrapping `=~` expression in
parenthesis.

Upstream-Status: Pending
Change-Id: If63a4b6981e5b562a9af7f2f8ac64947fad7f7b0
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

filipleple commented:
@SergiiDmytruk  so it looks like this branch fixes the capsule update issues. I've successfully tested on V540TU, V560TU, V540TNX - testing on V560TNX is in progress. Please rebase and fix the commit message names, so we can get to merging this

filipleple reviewed (dismissed):
also fixed [SMM BWP on NUC BOX](https://github.com/Dasharo/dasharo-issues/issues/1349) :fire:

krystian-hebel commented on src/soc/intel/common/block/fast_spi/mmap_boot.c:
```diff
@@ -22,28 +22,47 @@ enum window_type {
 	printk(BIOS_INFO, "SPI flash base=0x%lx, Host base=0x%lx, Size=0x%zx\n",
 	       flash_base, host_base, size);
+
+	real_dev_windows[real_dev_win_count++] = mmap_windows[type];
```
I would add a check against out-of-bounds writes, here and below. It will be easier in the future to see an error than accidentally overwriting data in neighboring variable, if it will ever be updated again.

In general, IMHO mixing iterated array indices and predefined positions is messy, but given how it is passed to `xlate_region_device_ro_init()` leaves us no choice, other than reworking the whole driver, which is beyond the scope of the task.

SergiiDmytruk replied:
https://github.com/Dasharo/coreboot/compare/703bb4ead23ab121aafc7c79c3aff118e128116e..e023501f15345e867265d6d9f7b762a4b2d10069#diff-f647b77e81916109378b68f46e5d9dfe122031abc093d31272b18a18dd074f32

krystian-hebel commented on src/soc/intel/common/block/fast_spi/mmap_boot.c:
```diff
@@ -22,28 +22,47 @@ enum window_type {
+	const struct region_device *rw_device = boot_device_rw();
+
+	xlate_window_init(&rw_window, rw_device, flash_base, size);
+	printk(BIOS_INFO, "Fast SPI RW Decode Window: SPI flash base=0x%zx, Size=0x%zx\n",
```
```suggestion
	printk(BIOS_INFO, "Fast SPI RW Decode Window: SPI flash base=0x%lx, Size=0x%zx\n",
```

SergiiDmytruk replied:
https://github.com/Dasharo/coreboot/compare/703bb4ead23ab121aafc7c79c3aff118e128116e..e023501f15345e867265d6d9f7b762a4b2d10069

krystian-hebel commented on src/drivers/efi/capsules.c:
```diff
@@ -792,8 +793,20 @@ static void enable_capsule_smi(void *unused)
 
-	ret = call_smm(APM_CNT_SMMSTORE, SMMSTORE_CMD_USE_FULL_FLASH,
-		       (void*)(uintptr_t)uefi_capsule_count);
+	/* SMM can occasionally ignore SMI calls, so retry several times on failure. */
```
Nit: technically, SMM can't ignore SMI because CPU is not in SMM at that point.

SergiiDmytruk replied:
https://github.com/Dasharo/coreboot/compare/703bb4ead23ab121aafc7c79c3aff118e128116e..e023501f15345e867265d6d9f7b762a4b2d10069#diff-16e632f09b8efab02dabf756cf47fe4e2b67fed218a821a178270840e5900a1b

krystian-hebel commented on payloads/external/edk2/Kconfig.dasharo:
```diff
@@ -4,7 +4,7 @@ config EDK2_REPOSITORY
 
 config EDK2_TAG_OR_REV
-	default "e4dea91cef0af9dbb05e7a407bb18ca18080cbaf"
+	default "e865b79c4e545c89362330792573fc70861b71d3"
```
edk2 will need a rebase, so this will change.

SergiiDmytruk replied:
https://github.com/Dasharo/coreboot/compare/703bb4ead23ab121aafc7c79c3aff118e128116e..e023501f15345e867265d6d9f7b762a4b2d10069#diff-f8ea1bb7fc113c7c1afaafc8cf4917177c1c7adad64c066b424c8fc88e88948f
